lib/ecto/query/planner.ex

```diff
@@ -952,6 +952,20 @@ defmodule Ecto.Query.Planner do
     {{:value, type}, [expr | fields], from}
   end
 
+  defp collect_fields({:coalesce, _, [left, right]} = expr, fields, from, query, take) do
+    {left_type, _, _} = collect_fields(left, fields, from, query, take)
+    {right_type, _, _} = collect_fields(right, fields, from, query, take)
+
+    type =
+      case {left_type, right_type} do
+        {_, {:value, type}} when type != :any -> {:value, type}
```

Comment:
Can we coalesce two different types?

In any case, I think this type could be written in two clauses?

    {{:value, :any}, type} -> type
    {type, _} -> type

Reply:
@josevalim I think the reason that we can't do that is because one of the types could be a value literal, like just `5` (from `coalesce(p.foo, 5)`), and I noticed that if I return that from this function that it will just use that value in all cases which would definitely be very wrong.. So its important here to only use the type returned by one of them *only if* `{:value, type}`. Of course correct me if I'm wrong in that, but that is what I noticed.

Reply:
I see! So let's add an assertion on this scenario on coalesce type as well.

Reply:
so asserting that in the case that the value is non-null and we are coalescing w/ a provided value `coalesce(p.some_not_null_field, 5)` that the value returned is not 5, but the value of the field?

Reply:
I also did not answer your question. In postgres, at least, you cannot coalesce values of different types. I have not quite done enough research to confirm this, but it appears to be dynamically defined for all types (or something along those lines) such that calling it with different types fails.

Reply:
> so asserting that in the case that the field is non-null and we are coalescing w/ a provided value coalesce(p.some_not_null_field, 5) that the value returned is not 5, but the value of the field?

:+1:

Reply:
Added the corresponding test.